src/bp/cluster.ts

```diff
@@ -68,5 +81,21 @@ export const setupMasterNode = (logger: sdk.Logger) => {
     }
   })
 
-  cluster.fork({ SERVER_ID: process.SERVER_ID })
+  const nluEnabled = !!config.modules.find(m => m.location.endsWith('nlu'))?.enabled
+  if (nluEnabled) {
+    spawnMLWorkers()
+  }
+  spawnWebWoker()
+}
+
+function spawnWebWoker() {
+  const { id } = cluster.fork({ SERVER_ID: process.SERVER_ID, WORKER_TYPE: <WorkerType>'WEB_WORKER' })
+  process.WEB_WORKER = id
+}
+
+function spawnMLWorkers() {
+  const maxMLWorkers = Math.max(os.cpus().length - 1, 1) // ncpus - webworker
+  const numMLWorkers = parseInt(process.env.NUM_ML_WORKERS || '', 10) || maxMLWorkers
+  process.ML_WORKERS = _.range(numMLWorkers).map(() => cluster.fork({ WORKER_TYPE: <WorkerType>'ML_WORKER' }).id)
+  debug(`Spawned ${numMLWorkers} machine learning workers`)
```

Comment:
the idea of soawning workers in the begining is to avoid spawning them afterwards as it takes quite a lot of precious runtime.

A single bot with n ctx will have 2*n +2 models to train and soon more (facets model + NDU) I thought that would be interresting to spawn as many workers as possible in a production environment but also to keep the ability to configure it.

I'd be willing to change, how would you "decide" when to spawn more workers?